### PR TITLE
Build: Speed up CI checks

### PR DESCRIPTION
**Pull request by gaetanmaisse**
Issue: CI is taking sooooo long to run the whole CircleCI workflow. Let's try to improve that!

## What I did

 - Replace handmade `repo-dirty-check` by `--frozen-lockfile` option when running `yarn install`: 1s -> 0s
 - ~Replace CircleCI `checkout` step by a custom one to only checkout 1 commit and avoid downloading the whole monorepo and its history: 20s -> 1s for each step~ -> Not working with PR from forks see https://github.com/storybookjs/storybook/pull/12315#discussion_r500537123
 - Tune executor usages to reduce time spent on bottleneck steps (with Large or XLarge executors) and use small executors for noncritical steps: not making a big diff alone 
 - Use the number of vcore minus 1 as concurrency level in `publish-to-verdaccio` script: ~3 min -> ~2min with a large executor 
 - Rework the steps: 
    - `build`: It's a merge of previous `install` and `build` steps as there is no real benefit to have 2 different steps and it avoids to persist and restore the workspace between theses steps. It _doesn't install dependencies from examples projects_ (i.e. delete `examples` directory before yarn install) to ensure the build can be done with deps listed in `@storybook` packages and not by chance thanks to examples deps. Built packages are then available in the workspace.
    - `install-examples-deps`: Install all dependencies required by `examples` and make them available in the workspace.
    - `install-e2e-deps`: Install only dependencies needed to run e2e tests and make them available in the workspace i.e. delete all sub-directory and so do a `yarn install` with only the root `package.json`. It has at least 2 benefits: fewer dependencies so faster bootstrap of e2e tests and, ensuring we have the required deps in root `package.json`.
    - All others jobs are still the same, only the previous jobs requirement has changed. All jobs are using `build` workspace and according to their need they are using either the `node_modules` from `install-examples-deps` or from `install-e2e-deps`

image

 
Everything combined workflow is running in ~15min (according to global workload) instead of ~25min 🎉 .

Let me know if you want me to add some doc somewhere or if some parts are not clear enough.

## How to test

- CircleCI and TeamCity should still be 🟢

### CODE REVIEW

**Comment by gaetanmaisse** [comment]
I guess all TeamCity jobs are failing because it somehow uses the workflow from `next` branch and the one of this branch 🤔

**Comment by gaetanmaisse** [comment]
> UI Tests Pending — 41 changes must be accepted as baselines.

Not sure about chromatic check, all stories reported as changes are the same as the one from next 🤔

**Review by shilman — APPROVED** [review]
This looks amazing @gaetanmaisse. Tremendous work! 💯 

I want @ndelangen to review before merging

**Review by shilman — COMMENTED** [review]
@gaetanmaisse @ndelangen any idea what's up with all the chromatic changes?

**Comment by ndelangen** [comment]
I can't explain these chromatic changes, I'll try merging in next

**Comment by ndelangen** [comment]
The tech/speed-up-ci-checks branch had 1 commit, and no parent commits.

I created a branch from next, and force merged tech/speed-up-ci-checks into it, resolving merge conflicts, but I'm very confused by the results, a lot of files changed I find puzzling.

**Comment by gaetanmaisse** [comment]
@ndelangen indeed that's weird, I still have `tech/speed-up-ci-checks` locally so I can try to update it with `next` and force push to see what happens.

**Comment by gaetanmaisse** [comment]
@ndelangen I rebased my branch and force-pushed it, the diff looks ok and it "only" has `+437 −360`

**Comment by yannbf** [comment]
The following changes [1](https://github.com/storybookjs/storybook/pull/12315/files#diff-9de053eff53b3da3ebb08e8a01781041R43) [2](https://github.com/storybookjs/storybook/pull/12315/files#diff-c61efba663a07d9aad178cf199022269R55) and [3](https://github.com/storybookjs/storybook/pull/12315/files#diff-6c085e0bdfbd605a8c503ba64854401fR62) will slightly impact new versions of `storybook/(aurelia,angular,addon-storyshots)`, should they be part of this PR as well?

**Comment by gaetanmaisse** [comment]
@yannbf I think it's ok for `storybook/(angular,addon-storyshots)` as I just fixed some missing deps/peerDep. For `aurelia` I set a proper version as `dev` tag is used to publish daily build so I'm not sure we should use it on our side. I put them in this PR because I was focusing on being able to build the packages and run tests without cheating with the dependencies of the whole monorepo.

I'm not strongly opinionated so I can't revert it if you think I should. What's your opinion on that?

**Comment by yannbf** [comment]
> @yannbf I think it's ok for `storybook/(angular,addon-storyshots)` as I just fixed some missing deps/peerDep. For `aurelia` I set a proper version as `dev` tag is used to publish daily build so I'm not sure we should use it on our side. I put them in this PR because I was focusing on being able to build the packages and run tests without cheating with the dependencies of the whole monorepo.
> 
> I'm not strongly opinionated so I can't revert it if you think I should. What's your opinion on that?

I'm not opinionated either, I was just thinking that this is a CI related PR which would/should not have production changes, and by touching these packages you might end up affecting their production code (as in a fix commit), even if it's just a peerDeps change. I'm not sure if it makes 100% sense, so don't feel blocked by this comment, I trust your senses! 👌

**Comment by tmeasday** [comment]
>> Replace CircleCI checkout step by a custom one to only checkout 1 commit and avoid downloading the whole monorepo and its history: 20s -> 1s for each step

> any idea what's up with all the chromatic changes?

These things are likely related 🤷‍♂️

**Comment by gaetanmaisse** [comment]
@tmeasday Thanks for pointing this out, I will take a look!

@yannbf If it's simple I will extract the few commits related to deps in another PR

**Comment by gaetanmaisse** [comment]
Thank you @tmeasday! I put back the "full" checkout check and chromatic is back to 🟢  🙇🏻 

@yannbf dep related commits were merged in another PR 🙂  

It looks good, latest runs (~ at the same time)
 - on next and another branch: `21min58`, `24min10`, `22min23`
 - on this branch: `15min43`, `14min47` 🎆 🎉 

@shilman @ndelangen I think this time we are good to go (except TeamCity 🤷🏻 )!

**Review by yannbf — APPROVED** [review]
Awesome work 🎉

**Review by shilman — APPROVED** [review]
🤯

**Comment by yannbf** [comment]
So what's left for this PR? Are the issues in TeamCity actual issues? I never know if I should trust those

**Comment by shilman** [comment]
We can ignore TeamCity, they are not required

File: .circleci/config.yml
Comment (ndelangen) [inline .circleci/config.yml:66]: Can you explain this a bit?

Comment (gaetanmaisse) [reply .circleci/config.yml]: Sure! I did this after having read https://discuss.circleci.com/t/checkout-depth/18699 and https://medium.com/equify-tech/from-5-minutes-to-5-seconds-customize-your-checkout-step-in-circleci-2-0-244d7b226d7d

But sadly I just reverted this commit as this custom checkout wasn't working with forks for at least 2 reasons:
 - the URL of the repo is the one from Storybook and not the fork (I didn't find a way to get the URL of the fork)
 - `CIRCLE_BRANCH` doesn't contain a branch name in case of a PR from a fork but something like `PR/1234`

**Review by tooppaaa — APPROVED** [review]

**Pull request merged**
